src/main/java/org/apache/sysml/hops/ConvolutionOp.java

```diff
@@ -0,0 +1,478 @@
+ * 
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.sysml.hops;
+
+import java.util.ArrayList;
+
+import org.apache.sysml.conf.ConfigurationManager;
+import org.apache.sysml.hops.Hop.MultiThreadedHop;
+import org.apache.sysml.hops.rewrite.HopRewriteUtils;
+import org.apache.sysml.lops.ConvolutionTransform;
+import org.apache.sysml.lops.Lop;
+import org.apache.sysml.lops.LopsException;
+import org.apache.sysml.lops.LopProperties.ExecType;
+import org.apache.sysml.parser.Expression.DataType;
+import org.apache.sysml.parser.Expression.ValueType;
+import org.apache.sysml.runtime.matrix.MatrixCharacteristics;
+import org.apache.sysml.runtime.util.ConvolutionUtils;
+
+public class ConvolutionOp extends Hop  implements MultiThreadedHop
+{	
+	private Hop.ConvOp op;
+	
+	long N = -1; long C = -1; long H = -1; long W = -1;
+	long K = -1; long R = -1; long S = -1;
+	long stride1 = -1;
+	long stride2 = -1;
+	long padding1 = -1;
+	long padding2 = -1;
+	long P = -1; long Q = -1;
+	
+	private int _maxNumThreads = -1; //-1 for unlimited
+
+	private ConvolutionOp() {
+		//default constructor for clone
+	}
+	
+	public ConvolutionOp(String l, DataType dt, ValueType vt, ConvOp o, Hop inp)
+	{
+		super(l, dt, vt);
+		op = o;
+		getInput().add(0, inp);
+		inp.getParent().add(this);
+		
+		//compute unknown dims and nnz
+		refreshSizeInformation();
+	}
+	
+	
+	public ConvolutionOp(String l, DataType dt, ValueType vt, ConvOp o, ArrayList<Hop> inp) 
+	{
+		super(l, dt, vt);
+		op = o;
+		
+		for( int i=0; i<inp.size(); i++ ) {
+			Hop in = inp.get(i);
+			getInput().add(i, in);
+			in.getParent().add(this);
+		}
+		
+		//compute unknown dims and nnz
+		refreshSizeInformation();
+	}
+
+	public ConvOp getOp()
+	{
+		return op;
+	}
+	
+	@Override
+	public String getOpString() {
+		return "" + HopsConv2Lops.get(op);
+	}
+
+	@Override
+	public Lop constructLops()
+		throws HopsException, LopsException 
+	{
+		//return already created lops
+		if( getLops() != null )
+			return getLops();
+		
+		ExecType et = optFindExecType();
+		ArrayList<Hop> inputs = getInput();
+		switch( op )
+		{
+			case IM2COL:
+			case RESHAPE_COL:
+			case ROTATE180:
+			case COL2IM:
+			case MAX_POOLING:
+			case MAX_POOLING_BACKWARD:
+			case DIRECT_CONV2D:
+			case DIRECT_CONV2D_BACKWARD_DATA:
+			case DIRECT_CONV2D_BACKWARD_FILTER:
+			{	
+				if( et == ExecType.CP )
+				{
+					setLops(constructConvolutionLops(et, inputs));
+					break;
+				}
+				else {
+					// TODO: Add support for SPARK/MR backends once we are happy with the performance of
+					// single node Lenet script. 
+					throw new HopsException("Unimplemented ConvolutionOp for execution type: " + et.name());
```

Comment:
To add Spark instructions (non-optimized version) for these functions, it might take a week or two.